TabMon/Sampler/CounterSampleDataTableHelper.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace TabMon.Sampler
+{
+    /// <summary>
+    /// Compares dictionaries comprised of datatable metadata.
+    /// </summary>
+    internal class CounterSampleDictionaryComparer : IEqualityComparer<IDictionary<string,string>>
+    {
+        private static IList<string> staticColumns = new[] { "Cluster", "Machine", "Source", "Category", "Instance", "Unit" };
+
+        public bool Equals(IDictionary<string,string> x, IDictionary<string,string> y)
+        {
+            foreach (var staticColumn in staticColumns)
+            {
+                if (x[staticColumn] != y[staticColumn])
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public int GetHashCode(IDictionary<string, string> dict)
+        {
+            StringBuilder sb = new StringBuilder();
+            foreach (var staticColumn in staticColumns)
+            {
+                sb.Append(dict[staticColumn]);
+            }
+            return sb.ToString().GetHashCode();
+        }
+    }
+
+    /// <summary>
+    /// Compresses a datatable based off the uniqueness of the metadata.
+    /// </summary>
+    public static class CounterSamplerResultHelper
+    {
+        private static readonly ILog Log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        /// <summary>
+        /// Compresses a data table into a semi-sparse data structure by evaluating the uniqueness of the metadata associated with each row.
+        /// </summary>
+        /// <param name="table">The original data table.</param>
+        /// <returns>The compressed data table.</returns>
+        public static DataTable CompressTable(DataTable table)
+        {
+            Dictionary<string, string> testString = new Dictionary<string, string>();
+            var distinctRowTypes = from DataRow dataRow in table.Rows
+                                  orderby dataRow["Instance"].ToString()
+                                  select new Dictionary<string, string>()
+                                  { { "Cluster", dataRow["Cluster"].ToString() },
+                                      { "Machine", dataRow["Machine"].ToString() },
+                                      { "Source", dataRow["Source"].ToString() },
+                                      { "Category", dataRow["Category"].ToString() },
+                                      { "Instance", dataRow["Instance"].ToString() },
+                                      { "Unit", dataRow["Unit"].ToString() }
+                                  };
+            var distinctRows = distinctRowTypes.Distinct(new CounterSampleDictionaryComparer()).ToList();
+
+            var compressedTable = table.Clone();
+
+            foreach (var distinctRow in distinctRows)
+            {
+                DataRow[] foundRows;
```

Comment:
Instead of forward-declaring this, let's join the declaration and assignment, e.g. line 78 can just be `DataRow[] foundRows = table.Select(selectString);`